File: Sample/Koudai/Server/release/PyScript/Action/action1441.py
import clr, sys
clr.AddReference('ZyGames.Framework.Common');
clr.AddReference('ZyGames.Framework');
clr.AddReference('ZyGames.Framework.Game');
clr.AddReference('ZyGames.Tianjiexing.Model');
clr.AddReference('ZyGames.Tianjiexing.BLL');
clr.AddReference('ZyGames.Tianjiexing.Lang');
clr.AddReference('ZyGames.Tianjiexing.Component');
from action import *
from System import *
from System.Collections.Generic import *
from ZyGames.Framework.Common.Log import *
from ZyGames.Tianjiexing.Model.ConfigModel import *
from ZyGames.Framework.Common import *
from ZyGames.Tianjiexing.Model import *
from ZyGames.Tianjiexing.BLL import *
from ZyGames.Tianjiexing.Lang import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Tianjiexing.Model.Config import *
from ZyGames.Tianjiexing.Model.Enum import *
from ZyGames.Tianjiexing.BLL.Base import *
from ZyGames.Tianjiexing.Component import *
from ZyGames.Framework.Common.Serialization import *

# 1441_佣兵升级界面接口
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.generalID = 0
        self.isUp = 0
        self.strUserItemID = []
        self.userID=''
        self.strUserCardNum = []

class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)
        self.userGeneral = None;
        self.upExperience = 0;
        self.nextLv = 0
        self.maxUserLv = 0;
        self.userLv = 0;
        self.costExperience = 0;
        self.currentGeneralExper = 0;  # 佣兵当前经验值
        self.experiencePercent = 0  # 升级经验百分比

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    if httpGet.Contains("GeneralID"):
        urlParam.generalID = httpGet.GetIntValue("GeneralID");
    else:
        urlParam.Result = False;
    return urlParam;

def GetItemBaseInfo(userID, userItemID):
    package = UserItemPackage.Get(userID);
    if package != None:
        useritem = package.ItemPackage.Find(lambda s:not s.IsRemove and s.UserItemID == userItemID);
        if useritem != None and useritem.Num > 0:
            return ConfigCacheSet[ItemBaseInfo]().FindKey(useritem.ItemID);
    return None;

def takeAction(urlParam, parent):
    actionResult = ActionResult();
    urlParam.userID = parent.Current.User.PersonalId;
    userId = parent.Current.User.PersonalId;
    contextUser = parent.Current.User;

    usergeneralList = GameDataCacheSet[UserGeneral]().FindAll(userId, lambda s:s.GeneralID != urlParam.generalID and s.GeneralType == GeneralType.YongBing,True);
    for item in usergeneralList:
        item.GeneralCard ='';
    userGeneral = GameDataCacheSet[UserGeneral]().FindKey(userId, urlParam.generalID);
    actionResult.userGeneral = userGeneral;
    if userGeneral != None:
        actionResult.nextLv = userGeneral.GeneralLv;
        if userGeneral.GeneralCard:
            urlParam.strUserItemID = userGeneral.GeneralCard.TrimEnd(',').Split(',');
    strList = "";
    for str in urlParam.strUserItemID:
        if(GetItemBaseInfo(userId,str)):
            strList += str + ",";
    if(strList!=""):
        strList = strList.TrimEnd(',');
        urlParam.strUserItemID = strList.Split(',');
        urlParam.strUserCardNum = userGeneral.GeneralCardNum.Split(',');
    else:
        urlParam.strUserItemID = [];
    generalInfo = ConfigCacheSet[GeneralInfo]().FindKey(urlParam.generalID);
    if generalInfo != None:
        urlParam.isUp = 1;       
    #generalEscalate = ConfigCacheSet[GeneralEscalateInfo]().FindKey(actionResult.nextLv, GeneralType.YongBing);
    #if generalEscalate is not None:
    #    actionResult.upExperience =generalEscalate.UpExperience;
    # 玩家最高等级
    actionResult.maxUserLv = ObjectExtend.ToShort(ConfigEnvSet.GetInt("User.CurrMaxLv"));
    # 玩家等级
    actionResult.userLv = contextUser.UserLv;
    # 升级经验百分比
    generalEscalate = ConfigCacheSet[GeneralEscalateInfo]().FindKey(actionResult.nextLv + 1, GeneralType.YongBing);
    if generalEscalate:
        actionResult.experiencePercent = generalEscalate.UpExperience

    return actionResult;

# 提升到最高等级
def improveLv(sumExperience, actionResult, tempUserLv):
    #while True:
    generalEscalate = ConfigCacheSet[GeneralEscalateInfo]().FindKey(actionResult.nextLv + 1, GeneralType.YongBing);
    if generalEscalate:
        if sumExperience >= generalEscalate.UpExperience:
            if actionResult.nextLv >= tempUserLv:
                return False;
            else:
                actionResult.nextLv += 1;
                sumExperience -= generalEscalate.UpExperience;
                actionResult.costExperience += generalEscalate.UpExperience;
                improveLv(sumExperience, actionResult, tempUserLv)
        else:
            return False;

def buildPacket(writer, urlParam, actionResult):
    userGeneral = actionResult.userGeneral
    actionResult.currentGeneralExper = userGeneral.CurrExperience;
    writer.PushIntoStack(len(urlParam.strUserItemID));
    index = 0;
    experienceCards = 0;
    for str in urlParam.strUserItemID:
        #if not str:
        #    itemInfo = None;
        #else:
        #    itemInfo = GetItemBaseInfo(urlParam.userID, str);
        itemInfo = None if str == None else GetItemBaseInfo(urlParam.userID, str);
        dsItem = DataStruct();
        dsItem.PushIntoStack(MathUtils.ToNotNullString(str));
        dsItem.PushIntoStack(0 if itemInfo == None else itemInfo.ItemID);
        dsItem.PushIntoStack('' if itemInfo == None else MathUtils.ToNotNullString(itemInfo.ItemName));
        dsItem.PushIntoStack('' if itemInfo == None else MathUtils.ToNotNullString(itemInfo.HeadID));
        count = MathUtils.ToInt(urlParam.strUserCardNum[index]);
        dsItem.PushIntoStack(count);
        # 经验卡
        experienceCards += itemInfo.EffectNum * count;
        writer.PushIntoStack(dsItem);
        index = index + 1;
    writer.PushIntoStack('' if userGeneral == None else MathUtils.ToNotNullString(userGeneral.GeneralName));
    writer.PushShortIntoStack(0 if userGeneral == None else MathUtils.ToShort(userGeneral.GeneralLv));
    # 当前经验 + 经验卡总经验
    sumExperience = actionResult.currentGeneralExper + experienceCards
    tempUserLv = actionResult.userLv * 3
    improveLv(sumExperience, actionResult, tempUserLv);
    if userGeneral.GeneralLv >= actionResult.maxUserLv or userGeneral.GeneralLv >= (actionResult.userLv * 3):
        actionResult.nextLv = 0;
    
    writer.PushShortIntoStack(MathUtils.ToShort(actionResult.nextLv));
    writer.PushIntoStack('' if userGeneral == None else MathUtils.ToNotNullString(userGeneral.PicturesID));
    writer.PushIntoStack(0 if userGeneral == None else userGeneral.LifeNum);
    diffExper = actionResult.costExperience - userGeneral.CurrExperience
    if actionResult.nextLv >= actionResult.userLv * 3 or actionResult.nextLv == 0:
        writer.PushIntoStack(diffExper);  # 实际获得经验
    else:
        writer.PushIntoStack(experienceCards);  # 获得经验卡经验
    writer.PushIntoStack(0 if userGeneral == None else userGeneral.AbilityID);
    writer.PushShortIntoStack(MathUtils.ToShort(urlParam.isUp));
    # 当前已有经验
    writer.PushIntoStack(0 if userGeneral == None else userGeneral.CurrExperience);
    
    # 当前已有经验占升级所需经验百分比
    if actionResult.experiencePercent > 0:
        proportion = userGeneral.CurrExperience / (actionResult.experiencePercent * 1.0);
        tempPro = '{0}%'.format(MathUtils.ToInt(proportion * 100));
        writer.PushIntoStack(tempPro);
    else:
        writer.PushIntoStack('0%');

    if actionResult.nextLv >= actionResult.userLv * 3 and experienceCards > diffExper:
        writer.PushIntoStack(1);
    else:
        writer.PushIntoStack(0);
    return True;

File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1441.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System.Collections.Generic;
using ZyGames.Framework.Game.Service;
using ZyGames.Framework.Game.Cache;
using ZyGames.Framework.Common;
using System.Data;
using ZyGames.Tianjiexing.Model;

namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 1441_佣兵升级界面接口
    /// </summary>
    public class Action1441 : BaseAction
    {
        private int generalID;
        private short nextLv;
        private short isUp;
        private string[] strUserItemID = new string[0];

        private UserGeneral userGeneral = null;

        public Action1441(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1441, httpGet)
        {

        }

        public override void BuildPacket()
        {
            this.PushIntoStack(strUserItemID.Length);
            foreach (var str in strUserItemID)
            {
                ItemBaseInfo itemInfo = string.IsNullOrEmpty(str) ? null : GetItemBaseInfo(ContextUser.UserID, str);
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack(str.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? 0 : itemInfo.ItemID);
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.ItemName.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.HeadID.ToNotNullString());
                PushIntoStack(dsItem);
            }
            this.PushIntoStack(userGeneral == null ? string.Empty : userGeneral.GeneralName.ToNotNullString());
            this.PushIntoStack(userGeneral == null ? (short)0 : userGeneral.GeneralLv);
            this.PushIntoStack((short)nextLv);
            this.PushIntoStack(userGeneral == null ? string.Empty : userGeneral.PicturesID.ToNotNullString());
            this.PushIntoStack(userGeneral == null ? 0 : userGeneral.LifeNum);
            this.PushIntoStack(userGeneral == null ? 0 : userGeneral.CurrExperience);
            this.PushIntoStack(userGeneral == null ? 0 : userGeneral.AbilityID);
            this.PushIntoStack((short)isUp);

        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetInt("GeneralID", ref generalID))
            {
                return true;
            }
            return false;
        }

        public override bool TakeAction()
        {
            var userGeneralList = new GameDataCacheSet<UserGeneral>().FindAll(ContextUser.UserID,
                                                                              s =>
                                                                              s.GeneralID != generalID &&
                                                                              s.GeneralType == GeneralType.YongBing && s.GeneralType != GeneralType.Soul);
            foreach (var general in userGeneralList)
            {
                general.GeneralCard = string.Empty;
            }
            userGeneral = new GameDataCacheSet<UserGeneral>().FindKey(ContextUser.UserID, generalID);
            if (userGeneral != null)
            {
                nextLv = MathUtils.Addition(userGeneral.GeneralLv, (short)1);
                if (!string.IsNullOrEmpty(userGeneral.GeneralCard))
                {
                    strUserItemID = userGeneral.GeneralCard.TrimEnd(',').Split(',');
                }
            }
            var generalInfo = new ConfigCacheSet<GeneralInfo>().FindKey(generalID);
            if (generalInfo != null)
            {
                isUp = 1;
            }
            return true;
        }

        /// <summary>
        /// 佣兵卡片物品
        /// </summary>
        /// <param name="userID"></param>
        /// <param name="userItemID"></param>
        /// <returns></returns>
        public static ItemBaseInfo GetItemBaseInfo(string userID, string userItemID)
        {
            var package = UserItemPackage.Get(userID);
            if (package != null)
            {
                var useritem = package.ItemPackage.Find(s => !s.IsRemove && s.UserItemID == userItemID);
                if (useritem != null)
                {
                    return new ConfigCacheSet<ItemBaseInfo>().FindKey(useritem.ItemID);
                }
            }
            return null;
        }
    }
}